File: 64.minimum-path-sum.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class Solution64
{
    public int MinPathSum(int[][] grid) {
        if(grid == null) return 0;
        int rows = grid.Length;
        if(rows ==0) return 0;
        int columns = grid[0].Length;
        if(columns ==0) return 0;

        int[,] pathMap = new int[rows, columns];

        for(int i =0; i < rows; i++)
        {
            pathMap[i, 0] = i ==0? grid[i][0]: grid[i][0]+ pathMap[i-1, 0];
        }
        
        for(int j =0; j < columns; j++)
        {
            pathMap[0, j] = j ==0? grid[0][j]: grid[0][j]+ pathMap[0, j-1];
        }
            
        for(int i =1; i < rows; i ++)
            for (int j =1; j < columns; j ++)
            {
                pathMap[i,j] =Math.Min(pathMap[i-1, j], pathMap[i,j-1]) + grid[i][j];
            }

        return pathMap[rows-1,columns-1];
    }
}

File: 64_minimum-path-sum.py
from typing import List


class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m =  len(grid)
        if m == 0: return 0
        n = len(grid[0])
        if n == 0: return 0
        # table = [[float('inf')]*n for i in range(m)]
        table = [[0]*n for i in range(m)]
        table[0][0] = grid[0][0]
        for j in range(1,n):
            table[0][j] = table[0][j-1] + grid[0][j]

        for i in range(1, m):
            table[i][0] = table[i-1][0] + grid[i][0]

            for j in range(1, n):
                table[i][j] = min(table[i-1][j], table[i][j-1]) + grid[i][j]
        
        return table[-1][-1]
instance = Solution()
print(instance.minPathSum([[1,3,1],[1,5,1],[4,2,1]]))

print(instance.minPathSum([[1,2,3],[4,5,6]]))
